separate acknowledgement from body, remove bibliography

krisbukovi opened the pull request:
Addressing issue #18 where acknowledgement tag is within the body tag, now acknowledgements are always moved outside of the body tag. The parsers were also rearranged as "html5lib" in bs4 does not remove certain tables like in the article referenced in this issue (2012MNRAS.419.3018C). When using this parser it only recognizes thead and tbody tags, all other nested tags are not recognized as children and are consequently not removed.  

This PR will also fix a problem where notes in a table footer are not extracted properly, an example of this is found in 2016ApJ...824…58D where a note in the Table 3 footer including "2MASS" is missing when html5lib is used. The reason the html5lib parser fails in extracting certain parts of the table footer is because of graphic tags that do not have closing tags, the parser tries to reconcile these and places closing graphic tags somewhere in or after the table footer and the graphic tag and its contents are then removed along with all or part of the table footer.

krisbukovi committed:
separate acknowledgement from body, remove bibliography, fix table removal after bs4 update

krisbukovi committed:
add unit test for notes in table footer

krisbukovi committed:
add unit test for parsers using _save_body_tag() function

krisbukovi committed:
rename unit test

krisbukovi committed:
add unit test for parsers that resolve namespaces

krisbukovi committed:
add unit tests for elsevier

krisbukovi committed:
add comment to unit test

krisbukovi committed:
add parser loop back to unit tests with the option to leave it empty to use default params

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -120,9 +136,12 @@ def test_that_we_can_extract_using_settings_template(self):
+                content = self.extractor.extract_multi_content(preferred_parser_names=(parser_name,))
+                self.assertEqual(rules.META_CONTENT['xml'].keys(), content.keys())
+        else:
             content = self.extractor.extract_multi_content(preferred_parser_names=(parser_name,))
```
`parser_name` is not defined on the second branch of the `if` statement

krisbukovi replied:
Fixed with most recent edits

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -52,10 +53,15 @@ def test_that_we_can_parse_the_xml_content(self):
+                self.assertEqual(journal_title, 'JOURNAL TITLE')
+        else:
+            self.extractor.parse_xml()
+            journal_title = self.extractor.extract_string('//journal-title')
```
I do not understand why we would need to have this `else` in every single unit test. I understand it could help us catch a mistake in the default argument of the `parse_xml` function, but we could have one single unit test to test that if it is necessary. Doing this in every unit test complicates the code and it won't get executed unless the developer sets `self.parsers` to None. Or maybe I am missing something else?

krisbukovi replied:
This isn't to help catch mistakes, it's so I can use the unit tests to debug when I'm making additions and/or edits to the code. Our unit tests don't say which parser is the one that is failing, and if there are multiple unit tests that have failed it can be frustrating/time consuming to figure out which parsers are the ones are causing the problem. When I am making modifications I don't necessarily want to see if all of the parsers work at that point in time, and the fact that you can't see which parsers have failed can lead to false assumptions that my decisions are based on during this process. To reiterate, I like having this option when I am in the process of modifying code, but in the future I would like to implement a more elegant approach with something like python 3's subtests or some sort of parameterization.

marblestation replied:
I see, then what you described is better achieved by setting `self.parsers = ('your parser of interest',)` temporarily (you revert it when you are done developing/debugging) instead of having all these `if` statements in each unit test.

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -311,11 +468,10 @@ def test_that_we_can_parse_the_xml_content(self):
-        for parser_name in ("html5lib", "html.parser", "lxml-html", "direct-lxml-html", "lxml-xml", "direct-lxml-xml",):
-            self.extractor.parse_xml(preferred_parser_names=(parser_name,))
-            journal_title = self.extractor.extract_string('//title')
+        self.extractor.parse_xml()
```
We are not going through the whole list of parsers here (in this whole class) while we were doing it before.

krisbukovi replied:
I added the loop back using TestTEIXMLExtractor's own self.parsers variable

marblestation commented on adsft/extraction.py:
```diff
@@ -662,10 +672,14 @@ def _parse_xml(self, parser_name):
 
+        # move acknowledgments after body (most likely only a minority of documents have this problem)
+        for e in parsed_xml.xpath(" | ".join(META_CONTENT['xml']['acknowledgements']['xpath'])):
+            self._append_tag_outside_parent(e)
```
Did we add a unit test covering this case? I may have missed it...

krisbukovi replied:
I had a unit test for acknowledgements but I added a more specific one for this issue

marblestation reviewed (changes requested):
I need more time to further review the changes but I wanted to provide already some feedback so that you can try to address or comment my suggestions, thanks!

krisbukovi committed:
add unit test, fix 2 unit test issues from comments

krisbukovi committed:
change name of unit test

krisbukovi committed:
remove maxDiff variable

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -259,12 +320,154 @@ def test_that_cdata_is_removed(self):
+                u"IV. SECTION IV THIS SECTION TESTS THAT COMMENTS ARE REMOVED. "
+                u"V. SECTION V THIS SECTION TESTS THAT CDATA IS REMOVED. "
+                u"Manual Entry 1 Manual Entry 2 TABLE I. TEXT a NOTES a TEXT"
+            )
```
In practice, the code here is doing the same as in `test_handling_of_parsers_that_remove_body_tag` but just with different parsers. We could merge these unit tests and consider all the parsers as proposed above.

krisbukovi replied:
Then we would have a test that checks if parsers are extracting everything correctly, but it doesn't give you any information as to why the test failed.

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -259,12 +320,154 @@ def test_that_cdata_is_removed(self):
+                    u"V. SECTION V THIS SECTION TESTS THAT CDATA IS REMOVED. " + s
+                )
+
+                content = self.extractor.extract_multi_content()
```
This illustrates the problem of all the duplicated code that we have in these `if` statements, this branch does not actually test for the acknowledgments. Apart from that, we are asserting again the body extraction for all parsers, so there is overlapping with the `test_handling_of_parsers_that_detect_namespaces` and `test_handling_of_parsers_that_remove_body_tag` unit tests.

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -259,12 +320,154 @@ def test_that_cdata_is_removed(self):
+                u"III. SECTION III THIS SECTION TESTS THAT THE TAIL IS PRESERVED . "
+                u"IV. SECTION IV THIS SECTION TESTS THAT COMMENTS ARE REMOVED. "
+                u"V. SECTION V THIS SECTION TESTS THAT CDATA IS REMOVED. " + s
+            )
```
This does not need to be necessarily limited to the parsers that remove the body tags but we could test all the parsers.

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -327,10 +532,9 @@ def test_that_we_can_extract_using_settings_template(self):
         full_text_content = self.extractor.open_xml()
-        for parser_name in ("html5lib", "html.parser", "lxml-html", "direct-lxml-html", "lxml-xml", "direct-lxml-xml",):
-            content = self.extractor.extract_multi_content(preferred_parser_names=(parser_name,))
+        content = self.extractor.extract_multi_content()
```
This is not iterating through all the parsers as we were before.

marblestation committed:
html5lib: Convert self closing xml tags to closing tags

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -355,10 +559,9 @@ def test_that_we_can_extract_acknowledgments(self):
         full_text_content = self.extractor.open_xml()
-        for parser_name in ("html5lib", "html.parser", "lxml-html", "direct-lxml-html", "lxml-xml", "direct-lxml-xml",):
-            content = self.extractor.extract_multi_content(preferred_parser_names=(parser_name,))
+        content = self.extractor.extract_multi_content()
```
This is not iterating through all the parsers as we were before.

marblestation committed:
Corrected tests removing html5lib exception plus other tweaks

- Recovered parser iterations when lost
- Merged unit tests that were asserting the body
- Corrected acknowledgement test

marblestation commented on adsft/extraction.py:
```diff
@@ -662,10 +672,14 @@ def _parse_xml(self, parser_name):
 
+        # move acknowledgments after body (most likely only a minority of documents have this problem)
+        for e in parsed_xml.xpath(" | ".join(META_CONTENT['xml']['acknowledgements']['xpath'])):
+            self._append_tag_outside_parent(e)
```
This should be moved to the end of the function, otherwise (for instance) it fails to remove formulas because the namespaces have not yet being removed/properly treated (depending on the parser).

marblestation commented on adsft/tests/test_extraction.py:
```diff
@@ -492,6 +742,69 @@ def test_that_we_can_extract_a_list_of_datasets(self):
+            for parser_name in ("lxml-xml",): # this is the only parser that works for this unit test
+                self.extractor.parse_xml(preferred_parser_names=(parser_name,))
+                section = self.extractor.extract_string('//body//section[@id="s3"]//para')
+                self.assertEqual(section, u'THIS SECTION TESTS THAT THE TAIL IS PRESERVED .')
```
This does not work for all the parsers because of some code in `_parse_xml` (file `adsft/extraction.py`) that needs to be moved to the end of the function (see comment above). One that is fixed, this unit test can go back to iterating through all the parsers.

marblestation committed:
Removal of elements should happen after namespaces have been corrected

marblestation committed:
Remove duplicated code in test, XML parser list in config

- Recovered original parsers order

NOTE:

If the developer needs to run the unit test imitating how the pipeline
works, meaning not iterating through all the parsers but only iterating
through the parsers until one succeeds, the developer can replace in
test_extraction.py:

```
self.preferred_parser_names = load_config().get('PREFERRED_XML_PARSER_NAMES')
```

by

```
self.preferred_parser_names = (None,)
```

marblestation reviewed (changes requested):
I run and debugged all the changes proposed in this PR on my machine. I identified the problems that I commented in this PR and I also implemented solutions, issuing a the PR https://github.com/krisbukovi/ADSfulltext/pull/5 that if merged, it will modify this PR.

My PR https://github.com/krisbukovi/ADSfulltext/pull/5 showcases how I think parsing bugs could be addressed and how I think we could achieve similar unit test coverage/functionalities without duplicating code. Please, let me know what you think and do not hesitate to share any concern that it are not yet addressed in that PR. Thanks!

krisbukovi committed:
Merge branch 'marblestation-bugfix_ack_graph_and_others' into separate_acknowledgements
